### ISSUE TITLE
JavaScript EventEmitter interface not working completely

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#4192]
It looks like only a few of the JavaScript driver's EventEmitter interface have ever been tested, specifically `on` and `addListener`. The rest look to be missing the `args` argument in the function signature, but then attempt to use that same variable in emitting the event. For clarity, here is a marked up version of the the `on` and `once` functions from [drivers/javascript/cursor.coffee](https://github.com/rethinkdb/rethinkdb/blob/80cbb95aa268fbc707cb4ae04209f875c92ea707/drivers/javascript/cursor.coffee) starting on line 268:

```coffeescript
    on: (args...) ->
        if not @emitter?
            @_makeEmitter()
            setImmediate => @_each @_eachCb
        @emitter.on(args...)

    once: -> # missing the args here
        if not @emitter?
            @_makeEmitter()
            setImmediate => @_each @_eachCb
        @emitter.once(args...)
```

Adding the `(args...)` to the method signatures fixes the problem, but we should do better and be passing the exact arguments that each function accepts.

**Comment by danielmewes** [comment]
Ouch.

@deontologician can you work on a fix for this please?

@larkost, @deontologician it would also be good to have a test for these functions for obvious reasons. Could you two coordinate to get something in?

**Comment by larkost** [comment]
I have the fix for the coffee script done already, and I actually found this by adding a couple of tests that will go in review shortly. I am going to expand that to cover all of the calls today. So I have this handled.

**Comment by deontologician** [comment]
Thanks again @larkost  :100:

**Comment by danielmewes** [comment]
Ah, perfect :+1:

**Comment by larkost** [comment]
The fix for the JavaSCript driver is in `next` with db9e640. The expanded tests will land with #4027.

**Comment by marshall007** [comment]
Sorry I'm a little late, but this could be generalized a bit which might make it less prone to similar mistakes in the future:

```coffee
class IterableResult
  # ...

[ 'addListener', 'on', 'once', <etc> ].forEach (method) ->
  IterableResult::[method] = ->
    if not @emitter?
      @_makeEmitter()
      setImmediate => @_each @_eachCb
    
    @emitter[method](arguments...)
```

I think the person who originally wrote this probably meant to use `arguments...` instead of `args...`.

**Comment by danielmewes** [comment]
@marshall007 Thanks for the tip. That's a neat piece of code.
Since we now have a working version already merged that's also more explicit, I think we won't change it again though.

**Comment by danielmewes** [comment]
@larkost could you also cherry-pick the fix into v2.0.x?

**Comment by larkost** [comment]
I added this to v2.0.x with 35e36ba.
